receiver/snmpreceiver/client.go

```diff
@@ -0,0 +1,512 @@
+					c.logger.Warn("Problem with closing connection while trying to reset it", zap.Error(err))
+				}
+				if err = c.Connect(); err != nil {
+					scraperErrors.AddPartial(len(oids), fmt.Errorf("problem with getting indexed data: problem connecting while trying to reset connection: %w", err))
+					return
+				}
+			}
+		} else if !walkFnCalled {
+			scraperErrors.AddPartial(1, fmt.Errorf("problem with getting indexed data: problem with SNMP WALK for OID '%v': Could not find any data for given OID", oid))
+		}
+	}
+}
+
+// chunkArray takes an initial array and splits it into a number of smaller
+// arrays of a given size.
+func chunkArray(initArray []string, chunkSize int) [][]string {
+	var chunkedArrays [][]string
+
+	for i := 0; i < len(initArray); i += chunkSize {
+		end := i + chunkSize
+
+		if end > len(initArray) {
+			end = len(initArray)
+		}
+
+		chunkedArrays = append(chunkedArrays, initArray[i:end])
+	}
+	return chunkedArrays
+}
+
+// convertSnmpPDUToSnmpData takes a piece of SnmpPDU data and converts it to the
+// client's snmpData type.
+func (c *snmpClient) convertSnmpPDUToSnmpData(pdu gosnmp.SnmpPDU) snmpData {
+	clientSNMPData := snmpData{
+		oid: pdu.Name,
+	}
+
+	// Condense gosnmp data types to our client's simplified data types
+	switch pdu.Type {
+	// Integer types
+	case gosnmp.Counter32:
+		fallthrough
+	case gosnmp.Gauge32:
+		fallthrough
+	case gosnmp.Uinteger32:
+		fallthrough
+	case gosnmp.TimeTicks:
+		fallthrough
+	case gosnmp.Integer:
+		value, err := c.toInt64(pdu.Name, pdu.Value)
+		if err != nil {
+			clientSNMPData.valueType = notSupportedVal
+			clientSNMPData.value = value
+			return clientSNMPData
+		}
+
+		clientSNMPData.valueType = integerVal
+		clientSNMPData.value = value
+		return clientSNMPData
+
+	// String types
+	case gosnmp.IPAddress:
+		fallthrough
+	case gosnmp.ObjectIdentifier:
+		fallthrough
+	case gosnmp.OctetString:
+		clientSNMPData.valueType = stringVal
+		clientSNMPData.value = toString(pdu.Value)
+		return clientSNMPData
+
+	// Float types
+	case gosnmp.OpaqueFloat:
+		fallthrough
+	case gosnmp.OpaqueDouble:
+		value, err := c.toFloat64(pdu.Name, pdu.Value)
+		if err != nil {
+			clientSNMPData.valueType = notSupportedVal
+			clientSNMPData.value = value
+			return clientSNMPData
+		}
+
+		clientSNMPData.valueType = floatVal
+		clientSNMPData.value = value
+		return clientSNMPData
+
+	// Not supported types either because gosnmp doesn't support them
+	// or they are a type that doesn't translate well to OTEL
+	case gosnmp.UnknownType:
+		fallthrough
+	case gosnmp.Counter64:
+		fallthrough
+	case gosnmp.NsapAddress:
+		fallthrough
+	case gosnmp.ObjectDescription:
+		fallthrough
+	case gosnmp.BitString:
+		fallthrough
+	case gosnmp.NoSuchObject:
+		fallthrough
+	case gosnmp.NoSuchInstance:
+		fallthrough
+	case gosnmp.EndOfMibView:
+		fallthrough
+	case gosnmp.Opaque:
+		fallthrough
+	case gosnmp.Null:
+		fallthrough
+	case gosnmp.Boolean:
+		fallthrough
```

Comment:
Why articulate all these cases? They will just hit the default condition either way.

Reply:
@djaglowski good callout. dumb mistake and will fix.